### PR TITLE
feat(ci-plus): make upload-artifacts have a dynamic matrix

### LINKED ISSUE #233: make 'plan' more like terraform
As discussed in https://github.com/axodotdev/cargo-dist/issues/178#issuecomment-1452020295, it would be more robust if our CI scripts used [dynamic matrices](https://docs.github.com/en/actions/using-jobs/using-a-matrix-for-your-jobs#example-using-contexts-to-create-matrices) that were defined by `cargo dist plan`.

Step 1 to this is to have `plan` always spit out the information we compute in ci.rs here:

https://github.com/axodotdev/cargo-dist/blob/4b3f86083fb008707665199611f697fc1aaa8d1d/cargo-dist/src/ci.rs#L80-L100

An agnostic version of this output could also be used to empower others to unilaterally bring up their own CI infra by parsing it and lowering it to their own system. This seems plausible because cargo-dist really doesn't care about the specifics of a machine -- it just cares about "windows, linux, mac" and requires that machine to have rustup and enough of a system to run the curl-sh/irm-iex installer to bootrstrap itself.

### PR DESCRIPTION
**Pull request by Gankra**
This also improves dist-manifest.json to include CI commands in its output, making it more like terraform

fixes #233

### CODE REVIEW

**Comment by Gankra** [comment]
Gonna be hammering https://github.com/Gankra/cargodisttest with this for a while to test/experiment.

File: cargo-dist/templates/ci/github_ci.yml.j2 [hunk 1]
```diff
@@ -110,39 +99,56 @@ jobs:
...
+        with:
+          name: artifacts
+          path: ${{ steps.cargo-dist.outputs.paths }}
 
-  # Mark the Github Release™ as a non-draft now that everything has succeeded!
   publish-release:
-    # Only run after all the other tasks, but it's ok if upload-artifacts was skipped
-    needs: [create-release, upload-artifacts]
-    if: ${{ always() && needs.create-release.result == 'success' && (needs.upload-artifacts.result == 'skipped' || needs.upload-artifacts.result == 'success') }}
+    # Create a Github Release with all the results once everything is done,
```
Comment (mistydemeo) [inline cargo-dist/templates/ci/github_ci.yml.j2:127]: No trademark? 😆

File: cargo-dist/templates/ci/github_ci.yml.j2 [hunk 2]
```diff
@@ -110,39 +99,56 @@ jobs:
...
+    # regardless of success. The release is only marked as non-draft if all tasks succeeded.
+    needs:
+      - plan
+      - upload-dist-artifacts
+    if: ${{ always() }}
```
Comment (mistydemeo) [inline cargo-dist/templates/ci/github_ci.yml.j2:132]: Curious what "if always" is doing here.

**Review by mistydemeo — APPROVED** [review]
Looking good to me.

File: cargo-dist/templates/ci/github_ci.yml.j2 (hunk 2 above)
Comment (Gankra) [reply cargo-dist/templates/ci/github_ci.yml.j2:132]: As soon as you want to permit some tasks to run if previous ones fail/cancel, you need to sprinkle this around a lot (can be useful for the publish-github-release step to make drafts still happen, although for now i've given up on pulling that functionality forward).

**Comment by Gankra** [comment]
Closing in favour of #378

**Pull request closed without merging**